org-code-javabuilder/lib/src/main/java/org/code/javabuilder/InputPoller.java

```diff
@@ -0,0 +1,44 @@
+package org.code.javabuilder;
+import com.amazonaws.services.sqs.AmazonSQS;
+import com.amazonaws.services.sqs.model.Message;
+
+import java.io.IOException;
+import java.util.List;
+
+// Details: https://examples.javacodegeeks.com/aws-sqs-polling-example-in-java/
+// https://docs.aws.amazon.com/code-samples/latest/catalog/javav2-sqs-src-main-java-com-example-sqs-SendMessages.java.html
+public class InputPoller extends Thread {
+  private final String queueUrl;
+  private final AmazonSQS sqsClient;
+  private final RuntimeIO runtimeIO;
+  private final JavaRunner javaRunner;
+  private final OutputHandler outputHandler;
+
+  public InputPoller(AmazonSQS sqsClient, String queueUrl, RuntimeIO runtimeIO, JavaRunner javaRunner, OutputHandler outputHandler) {
+    this.queueUrl = queueUrl;
+    this.sqsClient = sqsClient;
+    this.runtimeIO = runtimeIO;
+    this.javaRunner = javaRunner;
+    this.outputHandler = outputHandler;
+  }
+
+  public void run() {
+    while (javaRunner.isAlive()) {
+      List<Message> messages = sqsClient.receiveMessage(queueUrl).getMessages();
+
+      for (Message message : messages) {
+        try {
+          runtimeIO.passInputToProgram(message.getBody());
+        } catch (IOException e) {
+          outputHandler.sendMessage("There was an error passing input to your program." + e.getStackTrace());
+        }
+        sqsClient.deleteMessage(queueUrl, message.getReceiptHandle());
+      }
+      try {
+        Thread.sleep(400);
```

Comment:
I think we can prevent this busy waiting by polling the SQS client any time system.in.* is invoked. If anyone has tips on overriding system.in to add this functionality, please share!

Reply:
You can reset System.in to an InputStream with [System.setIn](https://docs.oracle.com/javase/7/docs/api/java/lang/System.html#setIn(java.io.InputStream). I think you could define a custom input stream that does some fancy stuff here, although I'm not sure how good of a pattern that would be. Another thing to think about is would there be a case when user input is provided in a way other than System.in? And will it always look like System.in when we are running student-provided code?

Reply:
I did something somewhat similar in my initial try at running student code for System out. [link](https://github.com/code-dot-org/java-ide/pull/7/files#diff-17458634177104ff6ee8dba2da91fd742a5e681ab48ca3d0c36b188e673481d1R63)